strawberry/permission.py

```diff
@@ -89,6 +111,85 @@ class AutoDirective:
 
         return self._schema_directive
 
+    @cached_property
+    def is_async(self) -> bool:
+        return iscoroutinefunction(self.has_permission)
+
+    def __and__(self, other: BasePermission):
+        return AndPermission([self, other])
+
+    def __or__(self, other: BasePermission):
+        return OrPermission([self, other])
+
+
+class CompositePermission(BasePermission, abc.ABC):
+    def __init__(self, child_permissions: List[BasePermission]):
+        self.child_permissions = child_permissions
+
+    def on_unauthorized(self, **kwargs: dict) -> Any:
+        failed_permissions = kwargs.get("failed_permissions", [])
+        for permission in failed_permissions:
+            permission.on_unauthorized()
+
+    @cached_property
+    def is_async(self) -> bool:
+        return any(x.is_async for x in self.child_permissions)
+
+
+class AndPermission(CompositePermission):
+    def has_permission(
+        self, source: Any, info: Info, **kwargs: dict
+    ) -> Union[bool, Awaitable[bool], (False, dict), Awaitable[(False, dict)]]:
+        if self.is_async:
+            return self._has_permission_async(source, info, **kwargs)
+
+        for permission in self.child_permissions:
+            if not permission.has_permission(source, info, **kwargs):
+                return False, {"failed_permissions": [permission]}
+        return True
+
+    async def _has_permission_async(
+        self, source: Any, info: Info, **kwargs: dict
+    ) -> Union[bool, (False, dict)]:
+        for permission in self.child_permissions:
+            if not await await_maybe(permission.has_permission(source, info, **kwargs)):
+                return False, {"failed_permissions": [permission]}
+        return True
+
+    def __and__(self, other: BasePermission):
+        self.child_permissions.append(other)
+        return self
+
+
+class OrPermission(CompositePermission):
+    def has_permission(
+        self, source: Any, info: Info, **kwargs: dict
+    ) -> Union[bool, Awaitable[bool], (False, dict), Awaitable[(False, dict)]]:
+        if self.is_async:
+            return self._has_permission_async(source, info, **kwargs)
+        failed_permissions = []
+        for permission in self.child_permissions:
+            if permission.has_permission(source, info, **kwargs):
+                return True
+            failed_permissions.append(permission)
+
+        return False, {"failed_permissions": failed_permissions}
```

Comment:
Probably get rid of dictionaries here, maybe make a dataclass instead 🤔

Reply:
The idea here was to remain as flexible as possible, allowing users to pass context in the future as well. With a dataclass we would be again very static about handling additional permission error context.

Reply:
Hm, I'm not sure how context would be used here, is it in the release noted (that I didn't read)? 😅

Reply:
It's internal for now, but it allows us to pass information to raise_exception from has_permission. We opted to separate exception raising from permission checking back when implementing this extension due to the fail silently feature.

Reply:
So basically the context contains "this permission failed out of the list of and / or permissions"